### PR TITLE
Add nn.LinearCrossEntropyLoss

### PR DESCRIPTION
**Pull request by pearu**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.13.0) (oldest at bottom):
* #176338
* __->__ #172446
* #172286

### CODE REVIEW

File: torch/nn/modules/loss.py [hunk 1]
```diff
@@ -1401,6 +1447,138 @@ def forward(self, input: Tensor, target: Tensor) -> Tensor:
...
+    num_classes: int
+    loss_dims: tuple[int, ...]
+    ignore_index: int
+    label_smoothing: float
+
+    def __init__(
+        self,
+        in_features: int,
+        num_classes: int,
+        loss_dims: tuple[int, ...] = (),
```
Comment (mikaylagawarecki) [inline torch/nn/modules/loss.py]:
```suggestion
        num_classes: int,
        *,
        loss_dims: tuple[int, ...] = (),
```

File: torch/nn/modules/loss.py [hunk 2]
```diff
@@ -59,6 +62,49 @@ def __init__(
...
+class _LinearLoss(_Loss):
+    def __init__(
+        self,
+        in_features: int,
+        out_features: int,
+        bias: bool = True,
+        device=None,
+        dtype=None,
+        size_average=None,
+        reduce=None,
```
Comment (mikaylagawarecki) [inline torch/nn/modules/loss.py]: nit: do we still need these legacy params here and below given that only the new `LinearCrossEntropyLoss` inherits from this?

File: torch/nn/modules/loss.py (hunk 1 above)
Comment (mikaylagawarecki) [inline torch/nn/modules/loss.py]: why did you choose to expose these as user facing arguments rather than deducing them based on the target shape?

"""
If containing class probabilities, the target must have
          shape :math:`(C)` or :math:`(N, C, d_1, d_2, ..., d_K)`
"""

File: torch/nn/modules/loss.py (hunk 2 above)
Comment (pearu) [reply torch/nn/modules/loss.py]: no, I'll remove these.

File: torch/nn/modules/loss.py (hunk 1 above)
Comment (pearu) [reply torch/nn/modules/loss.py]: `loss_dims` and `num_classes` are needed for initializing the linear module.
Notice that `out_features=num_classes * math.prod(loss_dims)`.

Comment (mikaylagawarecki) [reply torch/nn/modules/loss.py]: My bad, I missed that we wouldn't have the target till the first forward pass. To clarify, `loss_dims` contains the expected shapes of the relevant dimensions in the `target`, is that right?

 Is there a precedent for `loss_dims` that we can follow in any of the other loss functions? should we just use `out_features`?

Comment (pearu) [reply torch/nn/modules/loss.py]: Applied.

Comment (pearu) [reply torch/nn/modules/loss.py]: Right. Now using `out_features` instead of `loss_dims`.

**Comment by pearu** [comment]
@claude review this

**Comment by pearu** [comment]
@claude review this

**Comment by pearu** [comment]
@claude review this

**Comment by pearu** [comment]
@claude review this

**Comment by pearu** [comment]
@claude review this

**Comment by pearu** [comment]
@claude review this

**Comment by pearu** [comment]
@claude review this, assume that K-dimensional loss supports both class-indices and probablities targets.

**Comment by pearu** [comment]
@claude review this

**Comment by pearu** [comment]
@claude review this, re-check the review against the latest commit.

**Comment by slayton58** [comment]
@claude /pr-review

**Comment by slayton58** [comment]
@pearu I think code quality items 2 & 3 should be fixed. If you could take a look at the number of samples generated for testing too -- if that's what we need then that's what we need, but if we can cut it down for wallclock reasons, it may be worth exploring.

**Comment by pearu** [comment]
code quality (1): using keyword arguments now
code quality (2): ditto
code quality (3): fixed by code re-formatting
testing (1): the comment is cleaned up in #176338.
testing (2): reduced the number of samples by multiple of 2, the current set is now minimal to cover all code paths
testing (3): added the suggested comment

@claude /pr-review

**Review by slayton58 — APPROVED** [review]
Final changes look good - LGTM

**Comment by pearu** [comment]
@claude advice how to proceed with landing this reverted PR

**Comment by pearu** [comment]
Closing as replaced by https://github.com/pytorch/pytorch/pull/181573
